### ISSUE TITLE
Request support for MBox N4

### ISSUE THREAD

**Issue by n8886919** [MiraboxSpace/StreamDock-Device-SDK#15]
Does this sdk support for [MBox N4](http://e.tb.cn/h.Ta19BPixdePGJTk?tk=fvFQ3KeKOtR)?
It would be great if we could use its knob and touchscreen.

**Comment by malah-akh** [comment]
+1 and MBox N3

**Comment by Heart-State** [comment]
Will support later

**Comment by nowindi** [comment]
I could map All keys and movements in N3 but I could not set and image.

KEY_MAPPING = {
    1 : 11,
    2 : 12,
    3 : 13,
    4 : 14,
    5 : 15,
    6 : 6,
    7 : 7,
    8 : 8,
    9 : 9,
    10 : 10,
    11 : 1,
    12 : 2,
    13 : 3,
    14 : 4,
    15 : 5,
    37 : 7,
    48 : 8,
    49 : 9,
    53 : 10,
    51 : 11,
    52 : 12,
    80 : "big_left_turn",
    81 : "big_right_turn",
    144 : "1_left_turn",
    145 : "1_right_turn",
    96: "2_left_turn",
    97: "2_right_turn"

**Comment by Heart-State** [comment]
> I could map All keys and movements in N3 but I could not set and image.
> 
> KEY_MAPPING = { 1 : 11, 2 : 12, 3 : 13, 4 : 14, 5 : 15, 6 : 6, 7 : 7, 8 : 8, 9 : 9, 10 : 10, 11 : 1, 12 : 2, 13 : 3, 14 : 4, 15 : 5, 37 : 7, 48 : 8, 49 : 9, 53 : 10, 51 : 11, 52 : 12, 80 : "big_left_turn", 81 : "big_right_turn", 144 : "1_left_turn", 145 : "1_right_turn", 96: "2_left_turn", 97: "2_right_turn"

Yes, the setting button icon command of N3 is a little different from that of 293V2. We need to recompile a libtransport.so

**Comment by nowindi** [comment]
@Heart-State Do you have any idea of how to set the key images? is there any specific format and way to set them?

**Comment by Heart-State** [comment]
> @Heart-State Do you have any idea of how to set the key images? is there any specific format and way to set them?

![image](https://github.com/user-attachments/assets/45b6fdad-70d2-452c-874d-e8e68b733b09)
Is your image size correct?

**Comment by nowindi** [comment]
I am doing this

`           device.open()
            print("StreamDock open.")
            initial_brightness = 100
            device.wakeScreen()
            device.set_brightness(initial_brightness)


            key_image_path = "images/key_1.jpg" 
            device.clearAllIcon()
            key_index = 1
            device.cleaerIcon(1)
        
            device.set_key_image(key_index,key_image_path)
            logging.info("Image Set")
            device.set_brightness(100)
            device.clearAllIcon()
            device.refresh()
            logging.info("refresh")`
            
  It's not working the image is 64x64 (jpg and png) also I change between StreamDock293s and StreamDock293 in the device and change the value inside files StreamDock293s.py and StreamDock293.py from 100x100 to 64x64. 
  
  it is not working

**Comment by Heart-State** [comment]
> I am doing this
> 
> ```
>             device.open()
>             print("StreamDock open.")
>             initial_brightness = 100
>             device.wakeScreen()
>             device.set_brightness(initial_brightness)
>             key_image_path = "images/key_1.jpg" 
>             device.clearAllIcon()
>             key_index = 1
>             device.cleaerIcon(1)
>             device.set_key_image(key_index,key_image_path)
>             logging.info("Image Set")
>             device.set_brightness(100)
>             device.clearAllIcon()
>             device.refresh()
>             logging.info("refresh")
> ```
> 
> The image is 64x64 (jpg and png). Also I change between StreamDock293s and StreamDock293 in the deviceManager and change the value inside files StreamDock293s.py and StreamDock293.py from 100x100 to 64x64.
> 
> it is not working

The 293S stuff in the SDK seems to be incomplete, as if some parts have not been modified. The button icon size of 293S is 85*85 and the secondary screen on the right is 80*80. I am very sorry.

**Comment by nowindi** [comment]
I don't kwnow about refresh, cleaerIcon, clearAllIcon because I don't have any image in the streamDock Keys. the brightness set is working, I can see the light change.

**Comment by Heart-State** [comment]
> I don't kwnow about refresh, cleaerIcon, clearAllIcon because I don't have any image in the streamDock Keys. the brightness set is working, I can see the light change.

Have you tried the two image sizes 80×80 and 85×85?

**Comment by mad-bizarre** [comment]
Hi @Heart-State ! thanks for N4 support for linux! Very long time waited. It is normal that N4 is detected as two ?

Path: 1-2.4:1.0
Vendor ID: 26115
Product ID: 4103
Serial Number: 0300D0780619
Manufacturer: HOTSPOTEKUSB
Product: HOTSPOTEKUSB HID DEMO
Usage Page: 0
Usage: 0
Interface Number: 0
--------------------
Path: 1-2.4:1.1
Vendor ID: 26115
Product ID: 4103
Serial Number: 0300D0780619
Manufacturer: HOTSPOTEKUSB
Product: HOTSPOTEKUSB HID DEMO
Usage Page: 0
Usage: 0
Interface Number: 1
--------------------
Found 2 Stream Dock(s).

**Comment by mad-bizarre** [comment]
Also question - does API of N4 supports GIFs for buttons?

**Comment by Heart-State** [comment]
> Also question - does API of N4 supports GIFs for buttons?

Not supported yet,Follow-up can be supported

**Comment by Heart-State** [comment]
> Hi [@Heart-State](https://github.com/Heart-State) ! thanks for N4 support for linux! Very long time waited. It is normal that N4 is detected as two ?
> 
> ## Path: 1-2.4:1.0
> Vendor ID: 26115
> Product ID: 4103
> Serial Number: 0300D0780619
> Manufacturer: HOTSPOTEKUSB
> Product: HOTSPOTEKUSB HID DEMO
> Usage Page: 0
> Usage: 0
> Interface Number: 0
> ## Path: 1-2.4:1.1
> Vendor ID: 26115
> Product ID: 4103
> Serial Number: 0300D0780619
> Manufacturer: HOTSPOTEKUSB
> Product: HOTSPOTEKUSB HID DEMO
> Usage Page: 0
> Usage: 0
> Interface Number: 1
> Found 2 Stream Dock(s).

Yes, it is normal to get two devices, but on Linux, the Usage Page and the Usage value is 0 by default, and on Windows, we distinguish it by usage == 1. This is a problem, and we are currently trying to solve it.

**Comment by Heart-State** [comment]
> Also question - does API of N4 supports GIFs for buttons?
For GIF support, you need to process GIF into a frame-by-frame picture and then send it frame by frame, because the underlying device receives jpg and can't process GIF, and we also handle it in the software

**Comment by Heart-State** [comment]
MiraBox N4 is supported
